subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyHandlerProviderIntegrationTest.groovy

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.resolve
+
+import org.gradle.integtests.fixtures.AbstractHttpDependencyResolutionTest
+
+/**
+ * Tests covering the use of {@link org.gradle.api.provider.Provider} as an argument in the
+ * {@link org.gradle.api.artifacts.dsl.DependencyHandler} block.
+ */
+class DependencyHandlerProviderIntegrationTest extends AbstractHttpDependencyResolutionTest {
+
+    def "mutating the provider value before it's added resolves to the correct dependency"() {
+        when:
+        buildFile << """
+        configurations { conf }
+        
+        def lazyDep = objects.property(String).convention("org.mockito:mockito-core:1.8")
+        
+        dependencies {
+            conf lazyDep
+        }
+        
+        lazyDep.set("junit:junit:4.12")
+        
+        task checkDeps {
+            doLast {
+                def deps = configurations.conf.incoming.dependencies
+                assert deps.find { it instanceof ExternalDependency && it.group == 'junit' && it.name == 'junit' && it.version == '4.12' }
+                assert !deps.find { it instanceof ExternalDependency && it.group == 'org.mockito' && it.name == 'mockito-core' && it.version == '1.8'  }
+            }
+        }
+        """
+        then:
+        succeeds('checkDeps')
+    }
+
+    def "works correctly with up-to-date checking"() {
+        given:
+        mavenHttpRepo.module("group", "projectA", "1.1").publish()
+        mavenHttpRepo.module("group", "projectA", "1.2").publish()
+
+        when:
+        buildFile << """
+        repositories {
+            maven {
+                url = "${mavenRepo.uri}"
+            }
+        }
+        configurations { conf }
+        
+        dependencies {
+            conf provider { "group:projectA:\${property('project.version')}" }
+        }
+        
+        task retrieve (type: Sync) {
```

Comment:
:x: Tests should rather use `ResolveTestFixture` which is much stronger in its assertions.